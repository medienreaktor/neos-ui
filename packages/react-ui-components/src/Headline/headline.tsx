import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';

import {PickDefaultProps} from '../../types';

type HeadlineType = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

interface HeadlineTheme {
    readonly 'heading': string;
}

export interface HeadlineProps extends React.HTMLAttributes<HTMLHeadingElement> {
    /**
     * The contents to be rendered.
     */
    readonly children: React.ReactNode;

    /**
     * The semantic tag type of the headline.
     */
    readonly type: HeadlineType;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     * An optional css theme to be injected.
     */
    readonly theme?: HeadlineTheme;
}

type DefaultProps = PickDefaultProps<HeadlineProps, 'type'>;

export const defaultProps: DefaultProps = {
    type: 'h1',
};

class Headline extends PureComponent<HeadlineProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element {
        const {
            type,
            className,
            children,
            theme,
            ...rest
        } = this.props;
        const classNames = mergeClassNames(
            theme!.heading,
            className
        );

        switch (type) {
            case 'h1':
                return <h1 {...rest} className={classNames}>{children}</h1>;
            case 'h2':
                return <h2 {...rest} className={classNames}>{children}</h2>;
            case 'h3':
                return <h3 {...rest} className={classNames}>{children}</h3>;
            case 'h4':
                return <h4 {...rest} className={classNames}>{children}</h4>;
            case 'h5':
                return <h5 {...rest} className={classNames}>{children}</h5>;
            default:
                return <h6 {...rest} className={classNames}>{children}</h6>;
        }
    }
}

export default Headline;
