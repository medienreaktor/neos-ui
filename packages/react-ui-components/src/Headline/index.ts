import {themr} from '@friendsofreactjs/react-css-themr';
import identifiers from '../identifiers';

import Headline from './headline';
import style from './style.css';

export default themr(identifiers.headline, style)(Headline);
