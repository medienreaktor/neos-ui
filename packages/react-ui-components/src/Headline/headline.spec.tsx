import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import Headline, {defaultProps, HeadlineProps} from './headline';

describe('<Headline/>', () => {
    const props: HeadlineProps = {
        ...defaultProps,
        children: 'Foo children',
        theme: {
            'heading': 'headingClassName',
        }
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<Headline {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<Headline {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should allow the propagation of additional props to the wrapper.', () => {
        const wrapper = shallow(<Headline {...props} draggable={true}/>);
        expect(wrapper.prop('draggable')).toBe(true);
    });

    it('should render a "h1" node if no "type" prop was passed.', () => {
        const wrapper = shallow(<Headline {...props}/>);

        expect(wrapper.type()).toBe('h1');
    });

    it('should render a the appropriate node if a "type" prop was passed.', () => {
        expect(shallow(<Headline {...props} type="h2"/>).type()).toBe('h2');
        expect(shallow(<Headline {...props} type="h3"/>).type()).toBe('h3');
        expect(shallow(<Headline {...props} type="h4"/>).type()).toBe('h4');
        expect(shallow(<Headline {...props} type="h5"/>).type()).toBe('h5');
        expect(shallow(<Headline {...props} type="h6"/>).type()).toBe('h6');
    });
});
