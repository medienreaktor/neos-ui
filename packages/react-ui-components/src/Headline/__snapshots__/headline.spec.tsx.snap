// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Headline/> should render correctly. 1`] = `
<h1
  className="headingClassName"
>
  Foo children
</h1>
`;
